### PR TITLE
pd: try ask_split when ask_batch_split incompatible

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
When rolling update, pd version control will refuse AskBatchSplit request, thus split can't be performed during this process until all tikv bump to new version. We should use original AskSplit request when AskBatchSplit failed caused by compatibility. To let tikv know whether AskBatchSplit fail reason is compatibility or not, we introduce a new error type for ResponseHeader in https://github.com/pingcap/kvproto/pull/290

This PR mainly add code to try `ask_split` when `ask_batch_split` fails for incompatible, and other changes are all adding the old api back that https://github.com/tikv/tikv/pull/3064 deletes

## What are the type of the changes? (mandatory)
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
manual test
integration test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Refer to a related PR or issue link (optional)
https://github.com/tikv/tikv/pull/3064
https://github.com/pingcap/kvproto/pull/290

### CODE REVIEW

**Comment by Connor1996** [comment]
You can see here https://github.com/BusyJay/tikv/pull/3/files for the changes comparing to batch-split branch

**Comment by Connor1996** [comment]
PTAL @zhangjinpeng1987 @overvenus @BusyJay

File: src/pd/pd.rs [hunk 1]
```diff
@@ -222,8 +294,36 @@ impl<T: PdClient> Runner<T> {
...
+                    Err(Error::Incompatible) => {
+                        let tag = format!("[region {}] {}", region.id, peer.id);
+                        let task = Task::AskSplit {
+                            region,
+                            split_key: split_keys[0].clone(),
+                            peer,
+                            right_derive,
+                            callback,
+                        };
+                        if let Err(Stopped(t)) = scheduler.schedule(task) {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/pd/pd.rs]: `schedule()` only return `Err(Stopped())`  or `Ok()`, no need to handle other errors. It is a copy of `on_prepare_split_region()` for store.rs, and not handle either.

**Comment by Connor1996** [comment]
PTAL @zhangjinpeng1987 @overvenus @BusyJay

File: src/raftstore/coprocessor/split_observer.rs [hunk 2]
```diff
@@ -119,31 +119,48 @@ impl AdminObserver for SplitObserver {
...
+                let mut request = vec![req.take_split()];
+                if let Err(e) = self.on_split(ctx, &mut request) {
+                    error!(
+                        "[region {}] failed to handle split req: {:?}",
+                        ctx.region().get_id(),
+                        e
+                    );
+                    return Err(box_err!(e));
+                }
+                // self.on_split() makes sure request is not empty, or it will return error.
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/split_observer.rs]: Please move this comment to  L130.

File: src/raftstore/coprocessor/split_observer.rs [hunk 3]
```diff
@@ -232,8 +249,8 @@ mod test {
         assert!(!req.has_split(), "only split req should be handle.");
 
         req = new_split_request(b"test");
-        // Only batch split is supported.
-        assert!(observer.pre_propose_admin(&mut ctx, &mut req).is_err());
+        // For compatible issue, split should supported too.
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/split_observer.rs]: s/issue/reason

File: src/pd/pd.rs [hunk 4]
```diff
@@ -222,8 +294,40 @@ impl<T: PdClient> Runner<T> {
...
+                    // But if update time is long, it may cause large regions, so call ask_split instead.
+                    Err(Error::Incompatible) => {
+                        let tag = format!("[region {}] {}", region.id, peer.id);
+                        info!(
+                            "{} ask_batch_split is incompatible, use ask_split instead",
+                            tag
+                        );
+                        let task = Task::AskSplit {
+                            region,
+                            split_key: split_keys[0].clone(),
```
Comment (overvenus) [inline src/pd/pd.rs]: Can `clone` be skipped?

**Pull request merged**